#importando bibliotecas necessárias

In [422]:
import pandas as pd
import os

#criando caminho onde iremos armazenar os dados

In [423]:
localarq = ('/content/drive/MyDrive/Colab Notebooks/PÓS CIENT DADOS - XP/MÓDULO 2/')
arq = 'country_vaccinations.csv'
dados = os.path.join(localarq, arq)
dados

'/content/drive/MyDrive/Colab Notebooks/PÓS CIENT DADOS - XP/MÓDULO 2/country_vaccinations.csv'

#criação de um dataframe no pandas para ler o arquivo

In [424]:
dataframe = pd.read_csv(dados, sep=',', encoding='latin-1')

In [425]:
dataframe.head()

country iso_code        date  total_vaccinations  people_vaccinated  \
0  Afghanistan      AFG  2021-02-22                0.00               0.00   
1  Afghanistan      AFG  2021-02-23                 NaN                NaN   
2  Afghanistan      AFG  2021-02-24                 NaN                NaN   
3  Afghanistan      AFG  2021-02-25                 NaN                NaN   
4  Afghanistan      AFG  2021-02-26                 NaN                NaN   

   people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
0                      NaN                     NaN                 NaN   
1                      NaN                     NaN             1367.00   
2                      NaN                     NaN             1367.00   
3                      NaN                     NaN             1367.00   
4                      NaN                     NaN             1367.00   

   total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
0                            0.00                           0.00   
1                             NaN                            NaN   
2                             NaN                            NaN   
3                             NaN                            NaN   
4                             NaN                            NaN   

   people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
0                                  NaN                             NaN   
1                                  NaN                           34.00   
2                                  NaN                           34.00   
3                                  NaN                           34.00   
4                                  NaN                           34.00   

                                            vaccines  \
0  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
1  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
2  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
3  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
4  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   

                 source_name            source_website  
0  World Health Organization  https://covid19.who.int/  
1  World Health Organization  https://covid19.who.int/  
2  World Health Organization  https://covid19.who.int/  
3  World Health Organization  https://covid19.who.int/  
4  World Health Organization  https://covid19.who.int/

#analisando os dados

Sobre o dataset
Perguntas que precisamos responder (na carreira de um data analyst precisamos saber quais perguntas fazer)

1. Quais os países com maior número de pessoas imunizadas?
2. Qual vacina é a mais utilizada no mundo?
3. Qual país possui o maior número de vacinados por dia?
4. Qual a média de vacinação por país?

Por fim, exportaremos o arquivo.

In [426]:
#verificando as linhas
dataframe.shape[0]

86512

In [427]:
#verificando informações
dataframe.describe()

total_vaccinations  people_vaccinated  people_fully_vaccinated  \
count            43607.00           41294.00                 38802.00   
mean          45929644.64        17705077.79              14138299.85   
std          224600360.18        70787311.50              57139201.72   
min                  0.00               0.00                     1.00   
25%             526410.00          349464.25                243962.25   
50%            3590096.00         2187310.50               1722140.50   
75%           17012303.50         9152519.75               7559869.50   
max         3263129000.00      1275541000.00            1240777000.00   

       daily_vaccinations_raw  daily_vaccinations  \
count                35362.00            86213.00   
mean                270599.58           131305.49   
std                1212426.60           768238.77   
min                      0.00                0.00   
25%                   4668.00              900.00   
50%                  25309.00             7343.00   
75%                 123492.50            44098.00   
max               24741000.00         22424286.00   

       total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
count                        43607.00                       41294.00   
mean                            80.19                          40.93   
std                             67.91                          29.29   
min                              0.00                           0.00   
25%                             16.05                          11.37   
50%                             67.52                          41.44   
75%                            132.74                          67.91   
max                            345.37                         124.76   

       people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  
count                             38802.00                        86213.00  
mean                                 35.52                         3257.05  
std                                  28.38                         3934.31  
min                                   0.00                            0.00  
25%                                   7.02                          636.00  
50%                                  31.75                         2050.00  
75%                                  62.08                         4682.00  
max                                 122.37                       117497.00

In [428]:
dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 86512 entries, 0 to 86511
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   country                              86512 non-null  object 
 1   iso_code                             86512 non-null  object 
 2   date                                 86512 non-null  object 
 3   total_vaccinations                   43607 non-null  float64
 4   people_vaccinated                    41294 non-null  float64
 5   people_fully_vaccinated              38802 non-null  float64
 6   daily_vaccinations_raw               35362 non-null  float64
 7   daily_vaccinations                   86213 non-null  float64
 8   total_vaccinations_per_hundred       43607 non-null  float64
 9   people_vaccinated_per_hundred        41294 non-null  float64
 10  people_fully_vaccinated_per_hundred  38802 non-null  float64
 11  daily_vaccinations_per_milli

In [429]:
print(dataframe['country'].unique())

['Afghanistan' 'Albania' 'Algeria' 'Andorra' 'Angola' 'Anguilla'
 'Antigua and Barbuda' 'Argentina' 'Armenia' 'Aruba' 'Australia' 'Austria'
 'Azerbaijan' 'Bahamas' 'Bahrain' 'Bangladesh' 'Barbados' 'Belarus'
 'Belgium' 'Belize' 'Benin' 'Bermuda' 'Bhutan' 'Bolivia'
 'Bonaire Sint Eustatius and Saba' 'Bosnia and Herzegovina' 'Botswana'
 'Brazil' 'British Virgin Islands' 'Brunei' 'Bulgaria' 'Burkina Faso'
 'Burundi' 'Cambodia' 'Cameroon' 'Canada' 'Cape Verde' 'Cayman Islands'
 'Central African Republic' 'Chad' 'Chile' 'China' 'Colombia' 'Comoros'
 'Congo' 'Cook Islands' 'Costa Rica' "Cote d'Ivoire" 'Croatia' 'Cuba'
 'Curacao' 'Cyprus' 'Czechia' 'Democratic Republic of Congo' 'Denmark'
 'Djibouti' 'Dominica' 'Dominican Republic' 'Ecuador' 'Egypt'
 'El Salvador' 'England' 'Equatorial Guinea' 'Estonia' 'Eswatini'
 'Ethiopia' 'Faeroe Islands' 'Falkland Islands' 'Fiji' 'Finland' 'France'
 'French Polynesia' 'Gabon' 'Gambia' 'Georgia' 'Germany' 'Ghana'
 'Gibraltar' 'Greece' 'Greenland' 'Grenada

In [430]:
dataframe.query('country == "Brazil"').sort_values('date', ascending=False)[:5]

country iso_code        date  total_vaccinations  people_vaccinated  \
11085  Brazil      BRA  2022-03-29        413559595.00       181078067.00   
11084  Brazil      BRA  2022-03-28        412177025.00       180881628.00   
11083  Brazil      BRA  2022-03-27        411548202.00       180800024.00   
11082  Brazil      BRA  2022-03-26        411116197.00       180757995.00   
11081  Brazil      BRA  2022-03-25        410860641.00       180726798.00   

       people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
11085             160272858.00              1382570.00           656803.00   
11084             160008924.00               628823.00           546999.00   
11083             159842858.00               432005.00           587488.00   
11082             159756178.00               255556.00           548819.00   
11081             159673499.00               684805.00           596456.00   

       total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
11085                          193.26                          84.62   
11084                          192.61                          84.53   
11083                          192.32                          84.49   
11082                          192.12                          84.47   
11081                          192.00                          84.45   

       people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
11085                                74.90                         3069.00   
11084                                74.77                         2556.00   
11083                                74.70                         2745.00   
11082                                74.65                         2565.00   
11081                                74.62                         2787.00   

                                                vaccines  \
11085  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
11084  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
11083  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
11082  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
11081  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   

                                           source_name  \
11085  State governments via coronavirusbra1.github.io   
11084  State governments via coronavirusbra1.github.io   
11083  State governments via coronavirusbra1.github.io   
11082  State governments via coronavirusbra1.github.io   
11081  State governments via coronavirusbra1.github.io   

                          source_website  
11085  https://coronavirusbra1.github.io  
11084  https://coronavirusbra1.github.io  
11083  https://coronavirusbra1.github.io  
11082  https://coronavirusbra1.github.io  
11081  https://coronavirusbra1.github.io

In [431]:
dataframe['country'].nunique()

223

# '1. quais os países com maior número de pessoas imunizadas?'


In [432]:
#identificando a última data
ultima_data = dataframe['date'].max()
ultima_data

'2022-03-29'

In [433]:
#filtrando a última data
ultima_data = dataframe[dataframe.date == ultima_data]

In [434]:
#verificando a quantidade de linhas e colunas dentro de um dataset
ultima_data.shape

(54, 15)

ao retornar 54 linhas, a gente percebe que há falta de retorno, uma vez que são 223 países

##verificando a última data informada de cada país

In [435]:
dataframe.groupby('country')['date'].max().sort_values(ascending=False)

country
Lebanon                            2022-03-29
Lithuania                          2022-03-29
Kosovo                             2022-03-29
Kazakhstan                         2022-03-29
Japan                              2022-03-29
                                      ...    
Pitcairn                           2021-09-07
Turkmenistan                       2021-09-05
Bonaire Sint Eustatius and Saba    2021-09-01
Saint Helena                       2021-05-05
Falkland Islands                   2021-04-14
Name: date, Length: 223, dtype: object

##como conseguimos 223 retornos, igual o número de países, podemos continuar a análise

In [436]:
#buscando o total de vacinações
dataframe.groupby('country')['total_vaccinations'].max().sort_values(ascending=False)

country
China              3263129000.00
India              1834500657.00
United States       560181791.00
Brazil              413559595.00
Indonesia           377108938.00
                        ...     
Falkland Islands         4407.00
Montserrat               4211.00
Niue                     4161.00
Tokelau                  1936.00
Pitcairn                   94.00
Name: total_vaccinations, Length: 223, dtype: float64

In [437]:
#lista de top 10 países
imunizados_por_pais = dataframe.groupby('country')['total_vaccinations'].max().sort_values(ascending=False)
imunizados_por_pais

country
China              3263129000.00
India              1834500657.00
United States       560181791.00
Brazil              413559595.00
Indonesia           377108938.00
                        ...     
Falkland Islands         4407.00
Montserrat               4211.00
Niue                     4161.00
Tokelau                  1936.00
Pitcairn                   94.00
Name: total_vaccinations, Length: 223, dtype: float64

In [438]:
type(imunizados_por_pais)

pandas.core.series.Series

In [439]:
#como ele foi transformado em series ao fazer o filtro, ou seja, arrays, precisamos retornar a dataframe
imunizados = imunizados_por_pais.to_frame().reset_index()
imunizados

country  total_vaccinations
0               China       3263129000.00
1               India       1834500657.00
2       United States        560181791.00
3              Brazil        413559595.00
4           Indonesia        377108938.00
..                ...                 ...
218  Falkland Islands             4407.00
219        Montserrat             4211.00
220              Niue             4161.00
221           Tokelau             1936.00
222          Pitcairn               94.00

[223 rows x 2 columns]

In [440]:
#agora conseguimos verificar os top 10 países
imunizados = imunizados.sort_values('total_vaccinations', ascending=False)
imunizados

country  total_vaccinations
0               China       3263129000.00
1               India       1834500657.00
2       United States        560181791.00
3              Brazil        413559595.00
4           Indonesia        377108938.00
..                ...                 ...
218  Falkland Islands             4407.00
219        Montserrat             4211.00
220              Niue             4161.00
221           Tokelau             1936.00
222          Pitcairn               94.00

[223 rows x 2 columns]

In [441]:
#como o resultado saiu em exponencial, vamos converter o valor
def formatar_decimal(x):
    return f'{x:.2f}'

pd.options.display.float_format = formatar_decimal
imunizados10

country  total_vaccinations
0          China       3263129000.00
1          India       1834500657.00
2  United States        560181791.00
3         Brazil        413559595.00
4      Indonesia        377108938.00
5          Japan        254345587.00
6     Bangladesh        243642749.00
7       Pakistan        219368557.00
8        Vietnam        203144374.00
9         Mexico        191907868.00

# '2. Qual vacina é a mais utilizada no mundo?


In [442]:
vacinas_por_pais = dataframe[['country','date','vaccines']]
vacinas_por_pais

country        date  \
0      Afghanistan  2021-02-22   
1      Afghanistan  2021-02-23   
2      Afghanistan  2021-02-24   
3      Afghanistan  2021-02-25   
4      Afghanistan  2021-02-26   
...            ...         ...   
86507     Zimbabwe  2022-03-25   
86508     Zimbabwe  2022-03-26   
86509     Zimbabwe  2022-03-27   
86510     Zimbabwe  2022-03-28   
86511     Zimbabwe  2022-03-29   

                                                vaccines  
0      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
1      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
2      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
3      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
4      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
...                                                  ...  
86507  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  
86508  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  
86509  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  
86510  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  
86511  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  

[86512 rows x 3 columns]

In [443]:
#agora precisamos agrupar por país
vacinas_por_pais.groupby('vaccines')['country'].nunique().sort_values(ascending=False)

vaccines
Oxford/AstraZeneca                                                                                    20
Johnson&Johnson, Moderna, Oxford/AstraZeneca, Pfizer/BioNTech                                         17
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech                                                          15
Oxford/AstraZeneca, Pfizer/BioNTech                                                                   11
Johnson&Johnson, Moderna, Novavax, Oxford/AstraZeneca, Pfizer/BioNTech                                 8
                                                                                                      ..
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, Sinopharm/Beijing, Sinovac                               1
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, Sinovac, Sputnik Light, Sputnik V, ZF2001                1
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, Sinovac, Sputnik V                                       1
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, 

In [444]:
vacinas_pais = dataframe.groupby('country')['vaccines'].unique()
vacinas_pais = vacinas_pais.to_frame().reset_index()
vacinas_pais

country                                           vaccines
0          Afghanistan  [Johnson&Johnson, Oxford/AstraZeneca, Pfizer/B...
1              Albania  [Oxford/AstraZeneca, Pfizer/BioNTech, Sinovac,...
2              Algeria  [Oxford/AstraZeneca, Sinopharm/Beijing, Sinova...
3              Andorra     [Moderna, Oxford/AstraZeneca, Pfizer/BioNTech]
4               Angola                               [Oxford/AstraZeneca]
..                 ...                                                ...
218              Wales     [Moderna, Oxford/AstraZeneca, Pfizer/BioNTech]
219  Wallis and Futuna                                          [Moderna]
220              Yemen     [Johnson&Johnson, Oxford/AstraZeneca, Sinovac]
221             Zambia  [Johnson&Johnson, Oxford/AstraZeneca, Sinophar...
222           Zimbabwe  [Oxford/AstraZeneca, Sinopharm/Beijing, Sinova...

[223 rows x 2 columns]

# '3. Qual país possui o maior número de vacinados por dia?

In [445]:
vacinados_dia = dataframe.groupby('country')['daily_vaccinations'].max().sort_values(ascending=False).head()
vacinados_dia = vacinados_dia.to_frame().reset_index()
vacinados_dia

country  daily_vaccinations
0          China         22424286.00
1          India         10037995.00
2     Bangladesh          3758404.00
3  United States          3506960.00
4       Pakistan          2175773.00

In [446]:
linha = vacinados_dia['daily_vaccinations'].idxmax()
linha

0

In [447]:
print(f'O país com maior número de vacinados por dia é {vacinados_dia.iloc[linha,0]} com {vacinados_dia.iloc[linha,1]} vacinas.')

O país com maior número de vacinados por dia é China com 22424286.0 vacinas.


# '4. Qual a média de vacinação por dia por país?

In [448]:
media_pais = dataframe.groupby('country')['daily_vaccinations'].mean().sort_values(ascending=False)
media_pais = media_pais.to_frame().reset_index()
media_pais

country  daily_vaccinations
0                China          6930367.80
1                India          4175994.26
2        United States          1191726.82
3               Brazil           943528.75
4            Indonesia           846289.29
..                 ...                 ...
218  Wallis and Futuna               33.89
219            Tokelau               17.40
220               Niue               16.32
221         Montserrat               10.54
222           Pitcairn                0.51

[223 rows x 2 columns]

#criando a nova base de dados

##alterando formatação das bases

In [449]:
media_pais = media_pais.rename(columns = {'country':'pais', 'daily_vaccinations':'media_vacinacao'})
media_pais

pais  media_vacinacao
0                China       6930367.80
1                India       4175994.26
2        United States       1191726.82
3               Brazil        943528.75
4            Indonesia        846289.29
..                 ...              ...
218  Wallis and Futuna            33.89
219            Tokelau            17.40
220               Niue            16.32
221         Montserrat            10.54
222           Pitcairn             0.51

[223 rows x 2 columns]

In [450]:
vacinas_pais = vacinas_pais.rename(columns = {'country':'pais', 'vaccines':'vacina'})
vacinas_pais

pais                                             vacina
0          Afghanistan  [Johnson&Johnson, Oxford/AstraZeneca, Pfizer/B...
1              Albania  [Oxford/AstraZeneca, Pfizer/BioNTech, Sinovac,...
2              Algeria  [Oxford/AstraZeneca, Sinopharm/Beijing, Sinova...
3              Andorra     [Moderna, Oxford/AstraZeneca, Pfizer/BioNTech]
4               Angola                               [Oxford/AstraZeneca]
..                 ...                                                ...
218              Wales     [Moderna, Oxford/AstraZeneca, Pfizer/BioNTech]
219  Wallis and Futuna                                          [Moderna]
220              Yemen     [Johnson&Johnson, Oxford/AstraZeneca, Sinovac]
221             Zambia  [Johnson&Johnson, Oxford/AstraZeneca, Sinophar...
222           Zimbabwe  [Oxford/AstraZeneca, Sinopharm/Beijing, Sinova...

[223 rows x 2 columns]

In [451]:
imunizados = imunizados.rename(columns = {'country':'pais', 'total_vaccinations':'imunizados'})
imunizados

pais    imunizados
0               China 3263129000.00
1               India 1834500657.00
2       United States  560181791.00
3              Brazil  413559595.00
4           Indonesia  377108938.00
..                ...           ...
218  Falkland Islands       4407.00
219        Montserrat       4211.00
220              Niue       4161.00
221           Tokelau       1936.00
222          Pitcairn         94.00

[223 rows x 2 columns]

##exportando
Aqui vamos juntar e exportar
Para juntar, é importante que as colunas usadas para merge estejam nomedas igualmente

In [452]:
dados_tratados = pd.merge(imunizados, media_pais, how='inner')
dados_tratados

pais    imunizados  media_vacinacao
0               China 3263129000.00       6930367.80
1               India 1834500657.00       4175994.26
2       United States  560181791.00       1191726.82
3              Brazil  413559595.00        943528.75
4           Indonesia  377108938.00        846289.29
..                ...           ...              ...
218  Falkland Islands       4407.00            74.74
219        Montserrat       4211.00            10.54
220              Niue       4161.00            16.32
221           Tokelau       1936.00            17.40
222          Pitcairn         94.00             0.51

[223 rows x 3 columns]

In [453]:
dados_tratados = pd.merge(dados_tratados, vacinas_pais)
dados_tratados

pais    imunizados  media_vacinacao  \
0               China 3263129000.00       6930367.80   
1               India 1834500657.00       4175994.26   
2       United States  560181791.00       1191726.82   
3              Brazil  413559595.00        943528.75   
4           Indonesia  377108938.00        846289.29   
..                ...           ...              ...   
218  Falkland Islands       4407.00            74.74   
219        Montserrat       4211.00            10.54   
220              Niue       4161.00            16.32   
221           Tokelau       1936.00            17.40   
222          Pitcairn         94.00             0.51   

                                                vacina  
0    [CanSino, Sinopharm/Beijing, Sinopharm/Wuhan, ...  
1             [Covaxin, Oxford/AstraZeneca, Sputnik V]  
2          [Johnson&Johnson, Moderna, Pfizer/BioNTech]  
3    [Johnson&Johnson, Oxford/AstraZeneca, Pfizer/B...  
4    [Johnson&Johnson, Moderna, Novavax, Oxford/Ast...  
..                                                 ...  
218                               [Oxford/AstraZeneca]  
219                               [Oxford/AstraZeneca]  
220                                  [Pfizer/BioNTech]  
221                                  [Pfizer/BioNTech]  
222                               [Oxford/AstraZeneca]  

[223 rows x 4 columns]

In [ ]:
dados_tratados.to_csv('/content/drive/MyDrive/Colab Notebooks/PÓS CIENT DADOS - XP/MÓDULO 2/estatistica_pais.csv', index = True, sep = ';')